demonstrations/tutorial_gradient-methods.py

```diff
@@ -0,0 +1,1133 @@
+# a circuit that is parametrized by gates with a two-eigenvalue generator,
+# i.e. :math:`\mathrm{e}^{-\mathrm{i}\theta_j P}`. For instance :math:`P` could be a Pauli matrix.
+# Assume that the difference between the two eigenvalues is :math:`2r`. Then
+#
+# .. math::  \partial_j C = r \big[C(\theta + se_j) - C(\theta - se_j)\big], \quad s = \frac{\pi}{4r},
+#
+# where :math:`e_j` is the :math:`j`-th canonical unit vector.
+#
+
+
+@qml.qnode(dev, diff_method="parameter-shift")
+def cost(theta):
+    qfunc(theta)
+    return qml.expval(qml.PauliZ(1))
+
+
+grad_fn = qml.grad(cost)
+print(grad_fn(params))
+
+######################################################################
+# Note how this gives the exact same result as the ``backprop`` method, i.e. it is exact. On quantum
+# hardware, however, one would have to estimate the shifted cost function values with finite number of
+# shots, which leads to statistical errors.
+#
+
+######################################################################
+# References
+# ~~~~~~~~~~
+#
+# -  M. Schuld, et al., `Phys. Rev. A 99, 032331 <https://doi.org/10.1103/PhysRevA.99.032331>`__
+#    (2019)
+# -  Pennylane Demo: `Quantum gradients with
+#    backpropagation <https://pennylane.ai/qml/demos/tutorial_backprop.html>`__
+#
+
+######################################################################
+# Hadamard Test
+# =============
+#
+# When writing out the derivative of :math:`C` explicitly we can observe that it is equal to a very
+# similar expression as :math:`C` itself. For simplicity, assume here that
+# :math:`U(\theta) = V\mathrm{e}^{-\mathrm{i} \theta X}W` and :math:`\theta \in \mathbb{R}`. We then
+# get
+#
+# .. math::
+#
+#
+#        \partial_i C = \mathrm{i} \big[
+#            \langle \psi \vert \mathrm{e}^{\mathrm{i} \theta X} X \tilde{A} \mathrm{e}^{-\mathrm{i} \theta X} \vert \psi \rangle
+#            - \langle \psi \vert  \mathrm{e}^{\mathrm{i} \theta X} \tilde{A} X \mathrm{e}^{-\mathrm{i} \theta X} \vert \psi \rangle
+#        \big]
+#        = 2\mathrm{Im} \big[
+#            \langle \psi \vert \mathrm{e}^{\mathrm{i} \theta X} \tilde{A} X \mathrm{e}^{-\mathrm{i} \theta X} \vert \psi \rangle
+#        \big],
+#
+# where :math:`\vert \psi \rangle = W \vert 0 \rangle` and :math:`\tilde{A} = V^\dagger A V`. Hence,
+# we need to compute the imaginary part of the overlap of two states: \*
+# :math:`\vert \phi_1 \rangle = VX\mathrm{e}^{-\mathrm{i} \theta X}W \vert 0\rangle` \*
+# :math:`\vert \phi_2 \rangle = V\mathrm{e}^{-\mathrm{i} \theta X}W \vert 0\rangle`.
+#
+# This can be done via the Hadamard test. Note that the Hadamard test requires an ancilla qubit. For
+# our example circuit above this means we have to use a device which has three qubits.
+#
+# .. container::
+#
+#    ::
+#
+#       <img src="gradienthadamard.png">
+#       Image from <a href="http://frederikwil.de/slides/pas2020">frederikwil.de/slides/pas2020</a>.
+#
+
+dev_3qubits = qml.device("default.qubit", wires=3)
```

Comment:
```suggestion
dev = qml.device("default.qubit", wires=3)
```
I think it's ok just to redefine `dev` at this point.

Reply:
Unless the idea is to emphasize that you need now an extra qubit?

Reply:
I think, it's good to emphasize that one needs a different device than in the methods before. But I am also OK with calling both devices `dev`.